temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.common.converter;
+
+import com.google.common.base.Preconditions;
+import io.temporal.api.common.v1.Payload;
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.payload.codec.ChainCodec;
+import io.temporal.payload.codec.PayloadCodec;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+
+/**
+ * It's a delegating {@link DataConverter} implementation that wraps and chains both another {@link
+ * DataConverter} and several {@link PayloadCodec}s.
+ *
+ * <p>The underlying {@link DataConverter} is expected to be responsible for conversion between user
+ * objects and bytes represented as {@link Payloads}, while an underlying chain of codecs is
+ * responsible for a subsequent byte <-> byte manipulation such as encryption or compressions
+ */
+public class CodecDataConverter implements DataConverter, PayloadCodec {
+  private final DataConverter dataConverter;
+  private final PayloadCodec codec;
+
+  /**
+   * When serializing to Payloads:
+   *
+   * <ul>
+   *   <li>{@code dataConverter} is applied first, following by the chain of {@code codecs}.
+   *   <li>{@code codecs} are applied last to first meaning the earlier encoders wrap the later ones
```

Comment:
You might consider reversing these orders.

If I supply:

`new CodecDataConverter(converter, [compress, encrypt])`

That reads naturally to me as "during serialization, serialize, then compress, then encrypt", which is definitely the order you want. But with this setup, to do that properly, you have to write the list backwards which feels a little odd to me.

Reply:
I know what you are saying. 
But this is done this way to keep things consistent with Go and TypeScript. Also, GRPC interceptors are applied the same way, the tail is applied first.